File: network.cpp
#include <vector>
#include <iostream>
#include <cmath>

class Network {
public:
    std::vector<FCM*> neurons;
    std::vector<Synapse*> synapses;
    std::vector<double> input_currents;

    Network() {
        // Constructor to initialize vectors
    }

    int num_neurons() {
        return neurons.size();
    }

    int add_neuron() {
        FCM* neuron = new FCM();
        neurons.push_back(neuron);
        neuron->set_index(neurons.size() - 1); // Index for easy access to input_current array when synaptic current has to be inserted
        input_currents.push_back(0.0); // Initial input is zero
        std::cout << "Added neuron " << neuron->get_index() << std::endl;
        return neuron->get_index();
    }

    void connect(int pre_index, int post_index, double weight = 0.1, double delay = 1.0, double tau = 5.0, std::string syn_type = "excitatory") {
        if (pre_index == post_index) {
            std::cout << "Select two unique neurons to connect" << std::endl;
            return;
        }
        FCM* pre_neuron = neurons[pre_index];
        FCM* post_neuron = neurons[post_index];
        Synapse* synapse = new Synapse(pre_neuron, post_neuron, weight, delay, tau, syn_type);
        synapses.push_back(synapse);
        std::cout << "Connected pre-neuron " << pre_index << " to post neuron " << post_index << " and type = " << synapse->get_type() << std::endl;
    }

    std::vector<double> step(std::vector<double>& external_current, double t, double dt) {
        std::vector<double> Vm(num_neurons(), 0.0);

        // Update all neurons
        for (size_t n = 0; n < neurons.size(); n++) {
            FCM* neuron = neurons[n];
            neuron->step(input_currents[n] + external_current[n], dt);
            Vm[neuron->get_index()] = neuron->get_Vm(0);
        }

        // Set input currents for next step = 0
        std::fill(input_currents.begin(), input_currents.end(), 0.0);

        // Update all synapses and set synaptic currents for next time step
        for (auto& synapse : synapses) {
            synapse->step(t, dt);
            input_currents[synapse->get_post_neuron()->get_index()] += synapse->get_current() * 1e-6;
        }

        return Vm;
    }
};

File: network.py
import numpy as np
from FCM_intra import FCM
from synapse import Synapse

class Network:
    def __init__(self):
        self.neurons = []
        self.synapses = []
        # To keep track of synaptic and external currents
        self.input_currents = []

    def add_neuron(self):
        neuron = FCM()
        self.neurons.append(neuron)
        neuron.set_index(len(self.neurons) - 1) # index for easy access to input_current array when synaptic current has to be inserted
        input_currents.append(0) # initial input is just zero
        return neuron.index

    def connect(self, pre_index, post_index, weight=0.1, delay=1.0, tau=5.0, syn_type="excitatory"):
        if pre_index == post_index:
            print("Select two unique neurons to connect")
            return 
        pre_neuron = self.neurons[pre_index]
        post_neuron = self.neurons[post_index]
        synapse = Synapse(pre_neuron, post_neuron, weight, delay, tau, syn_type)
        self.synapses.append(synapse)

    def step(self, external_current, t, dt):
        """
        Simulate the network.
        
        Args:
        - external_current: array with external input current for neurons (given in terms of uA)
        - t: Time now (ms)
        - dt: Time step (ms).
        """

        # Update all neurons
        for n in range(len(self.neurons)):
            neuron = self.neurons[n]
            neuron.step(curr_input = self.input_currents[n][i], dt)
        
        # Update all synapses
        for synapse in self.synapses:
            synapse.step(t, dt)

        # Update input currents for next step


# Example usage
if __name__ == "__main__":
    # Create a network
    network = Network()

    # Add neurons to the network
    network.add_neuron(neuron1)
    network.add_neuron(neuron2)

    # Connect neurons with synapses
    network.connect(0, 1, weight=0.5, delay=1.0, tau=10.0, syn_type="excitatory")

    # Simulate the network
    network.run(sim_end=100, dt=0.1)

